### PR TITLE
393 add Keep Largest Connected Component transform

### LINKED ISSUE #393: add keep_largest_connected_component transform
**Is your feature request related to a problem? Please describe.**
"keep_largest_connected_component" is very useful a transform in medical tasks.
Need to add it to MONAI.

### PR DESCRIPTION
**Pull request by YuanTingHsieh**
Fixes #393 .

### Description
Add KeepLargestConnectedComponent transform
Array/Dictionary version along with tests.

### Status
**Ready**

### Types of changes

- [] Bug fix (non-breaking change which fixes an issue)
- [x] Breaking change (fix or new feature that would cause existing functionality to change)
- [x] New tests added to cover the changes
- [x] Docstrings/Documentation updated

### CODE REVIEW

File: monai/transforms/utility/array.py [hunk 1]
```diff
@@ -251,3 +252,44 @@ def __init__(self, delay_time=0.0):
...
+        self.is_independent = is_independent
+
+    def __call__(self, img, applied_labels):
+        """
+        Args:
+            img (ndarray): shape must be (num_channels, H, W[, D]).
+            applied_labels (list or tuple of int): label number list for applying the connected component on.
+                The label that is not in this list will remain unchanged.
+        """
+        output = np.array(img)
```
Comment (Nic-Ma) [inline monai/transforms/utility/array.py:280]: You can edit on `img` directly as the img will be useless after this transform in the transform chain.

File: monai/transforms/utility/array.py [hunk 2]
```diff
@@ -251,3 +252,44 @@ def __init__(self, delay_time=0.0):
...
+class KeepLargestConnectedComponent(Transform):
+    """
+    Keeps only the largest connected component.
+
+    Value 0 will be treated as background.
+    """
+    def __init__(self, is_independent=True):
+        """
+        Args:
+            is_independent (bool): consider several labels as a whole or independent, default is `True`.
```
Comment (Nic-Ma) [inline monai/transforms/utility/array.py:266]: Maybe `independent` is enough for the arg name?

File: monai/transforms/utility/array.py [hunk 3]
```diff
@@ -251,3 +252,44 @@ def __init__(self, delay_time=0.0):
...
+                Example use case would be segment label 1 is liver and label 2 is liver tumor, in that case
+                you want this "is_independent" to be specified as False.
+        """
+        super().__init__()
+        self.is_independent = is_independent
+
+    def __call__(self, img, applied_labels):
```
Comment (Nic-Ma) [inline monai/transforms/utility/array.py:273]: We may not run this transform on `label` data only, so maybe "applied_values" and assert it must be `int`?
And I think it's better to move this arg to "__init__()".

File: monai/transforms/utils.py [hunk 4]
```diff
@@ -392,3 +393,20 @@ def generate_spatial_bounding_box(img, select_fn=lambda x: x > 0, channel_indexe
...
+
+
+def get_largest_connected_component_mask(img):
+    """
+    Gets the largest connected component mask of an image.
+
+    Args:
+        img (ndarrary): Image to get largest connected component from.
+    """
+    labels = measure.label(img > 0)
```
Comment (Nic-Ma) [inline monai/transforms/utils.py:405]: `measure.label` has an important parameter: `connectivity`, I think we should support to config it. It affects the connected area a lot. What do you think?

File: tests/test_keep_largest_connected_component.py [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+
+import unittest
+
+import numpy as np
+from parameterized import parameterized
+
+from monai.transforms import KeepLargestConnectedComponent
+
+grid_1 = np.array([[[0, 0, 1, 0, 0], [0, 2, 1, 1, 1], [1, 2, 1, 0, 0], [1, 2, 0, 1, 0], [2, 2, 0, 0, 2]]])
+grid_2 = np.array([[[0, 0, 0, 0, 1], [0, 0, 1, 1, 1], [1, 0, 1, 1, 2], [1, 0, 1, 2, 2], [0, 0, 0, 0, 1]]])
```
Comment (Nic-Ma) [inline tests/test_keep_largest_connected_component.py:20]: May I know what's the difference between tests for `grid_1` and `grid_2`?

File: tests/test_keep_largest_connected_componentd.py [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+    {"keys": ["img"], "is_independent": False, "applied_labels": [1, 2]},
+    grid_2,
+    np.array([[[0, 0, 0, 0, 1], [0, 0, 1, 1, 1], [0, 0, 1, 1, 2], [0, 0, 1, 2, 2], [0, 0, 0, 0, 1]]]),
+]
+
+VALID_CASES = [TEST_CASE_1, TEST_CASE_2, TEST_CASE_3, TEST_CASE_4,
+               TEST_CASE_5, TEST_CASE_6, TEST_CASE_7]
+
+
+class TestKeepLargestConnectedComponent(unittest.TestCase):
```
Comment (Nic-Ma) [inline tests/test_keep_largest_connected_componentd.py:75]: class name should be `TestKeepLargestConnectedComponentd`.

**Comment by Nic-Ma** [comment]
Hi @YuanTingHsieh ,

And according to our offline discussion today, this transform is only used to handle model output.
Please move to **/transforms/post/** and change to handle `PyTorch Tensor` data instead of numpy.

Thanks.

**Comment by Nic-Ma** [comment]
According to @tvercaut 's suggestion, let's convert Tensor to numpy array before calling  `skimage.measure.label()`, then convert back to Tensor for the current version.
And may replace this part with Kornia in a future version.

Thanks.

File: monai/transforms/utils.py (hunk 4 above)
Comment (YuanTingHsieh) [reply monai/transforms/utils.py:405]: sure, will do

File: tests/test_keep_largest_connected_component.py (hunk 5 above)
Comment (YuanTingHsieh) [reply tests/test_keep_largest_connected_component.py:20]: The first one looks like
0, 0, 1, 0, 0
0, 2, 1, 1, 1
1, 2, 1, 0, 0
1, 2, 0, 1, 0
2, 2, 0, 0, 2

Second one looks like
0, 0, 0, 0, 1
0, 0, 1, 1, 1
1, 0, 1, 1, 2
1, 0, 1, 2, 2
0 ,0 ,0 ,0 ,1

The second case have 1's that are not reachable under all connectivity or independent.

File: monai/transforms/utility/array.py [hunk 7]
```diff
@@ -251,3 +252,44 @@ def __init__(self, delay_time=0.0):
...
+                foreground = (img == i).astype(np.uint8)
+                mask = get_largest_connected_component_mask(foreground)
+                output[img == i] = 0
+                output += mask * i
+        else:
+            foreground = np.zeros(shape=img.shape, dtype=img.dtype)
+            for i in applied_labels:
+                foreground += (img == i).astype(np.uint8)
+            mask = get_largest_connected_component_mask(foreground)
+            output = mask * img
```
[the remark this answers is not in the document]
Comment (YuanTingHsieh) [reply monai/transforms/utility/array.py:294]: yes, you are correct.
will fix this to match the description

File: monai/transforms/utility/array.py [hunk 8]
```diff
@@ -251,3 +252,44 @@ def __init__(self, delay_time=0.0):
     def __call__(self, img, delay_time=None):
         time.sleep(self.delay_time if delay_time is None else delay_time)
         return img
+
+
+class KeepLargestConnectedComponent(Transform):
+    """
+    Keeps only the largest connected component.
```
[the remark this answers is not in the document]
Comment (YuanTingHsieh) [reply monai/transforms/utility/array.py:259]: will do

**Comment by YuanTingHsieh** [comment]
Thanks for the detailed review and feedback.
I have fixed those issues.

One thing here is to support the background pixel.
You can take a look at testing case 11 to see if that match what you original think.

File: monai/transforms/post/array.py [hunk 9]
```diff
@@ -46,3 +49,93 @@ def __call__(self, img, to_onehot=None, num_classes=None):
...
+        Args:
+            img: shape must be (num_channels, spatial_dim1[, spatial_dim2, ...]).
+
+        Returns:
+            A PyTorch Tensor with shape (1, spatial_dim1[, spatial_dim2, ...]).
+        """
+        if img.size()[0] == 1:
+            img = torch.squeeze(img, dim=0)
+        else:
+            img = torch.argmax(img, dim=0)
```
Comment (Nic-Ma) [inline monai/transforms/post/array.py:127]: Let's don't do `Argmax`, it's the work of another post transform `ConvertForMetrics`, will do later.
And we will try to support One-Hot format data(multi-labels task) after this PR merged, which doesn't expect Argmax operation for channels.

Comment (YuanTingHsieh) [reply monai/transforms/post/array.py:127]: If we don’t do argmax here, we don’t support multichannel input in this transform then. 
Because if input is a multichannel tensor that represents the output, I think each channel will corresponds to one label. And usually they are probability

If we don’t want to do extra thing in this transform, I think we can add in docstring saying we just support one channel where the pixel value in it is the prediction label.

Or you think there might be cases where each channel is having their own prediction label of multiple classes?

Comment (Nic-Ma) [reply monai/transforms/post/array.py:127]: 1. If the model output is for "multi-labels" task and you hard-code to do `Argmax` will cause wrong output.
2. And even if the model output only has 1 channel but not rounds to 0 or 1, this transform also can't handle it. Need to add sigmoid first and round values by (> 0.5).

So in order to handle all kinds of model outputs, I am developing a separate post-transform, and can make a **post-transform chain** with this KeepLargestConnectedComponnent and SplitChannel transforms. The signature of that transform looks like:
```py
class ConvertForMetrics(Transform):
    def __init__(
        self,
        y_pred,
        y,
        add_sigmoid=False,
        add_argmax=False,
        round_values=False,
        to_onehot_y_pred=False,
        to_onehot_y=False,
        logit_thresh=0.5
    )
```
 Let's keep this PR simple to support only 1 channel first, then let's work together to submit another PR to support multi-channels case for this issue: https://github.com/Project-MONAI/MONAI/issues/420
What do you think?

Thanks.

File: monai/transforms/post/dictionary.py [hunk 10]
```diff
@@ -56,4 +56,35 @@ def __call__(self, data):
...
+                Accepted values are ranging from  1 to input.ndim. If ``None``, a full
+                connectivity of ``input.ndim`` is used.
+        """
+        super().__init__(keys)
+        self.converter = KeepLargestConnectedComponent(applied_values, independent, background, connectivity)
+
+    def __call__(self, data):
+        d = dict(data)
+        for idx, key in enumerate(self.keys):
+            d[key] = self.converter(d[key])
```
Comment (Nic-Ma) [inline monai/transforms/post/dictionary.py:85]: BTW, in order to keep the original model output for other post transforms or some additional computation, let's add a new key to store the data in `key_{postfix}`.
Refer to: https://github.com/Nic-Ma/MONAI/blob/417-add-ConvertForMetrics-post-transform/monai/transforms/post/dictionary.py#L112

Thanks.

File: monai/transforms/post/array.py (hunk 9 above)
Comment (YuanTingHsieh) [reply monai/transforms/post/array.py:127]: I will go with the first, which is remove argmax and add docstring "Expected input data should have only 1 channel and the values correspond to expected labels."

File: monai/transforms/utils.py [hunk 11]
```diff
@@ -392,3 +394,24 @@ def generate_spatial_bounding_box(img, select_fn=lambda x: x > 0, channel_indexe
...
+    Gets the largest connected component mask of an image.
+
+    Args:
+        img: Image to get largest connected component from.
+        connectivity (int): Maximum number of orthogonal hops to consider a pixel/voxel as a neighbor.
+            Accepted values are ranging from  1 to input.ndim. If ``None``, a full
+            connectivity of ``input.ndim`` is used.
+    """
+    img_arr = img.detach().cpu().numpy()
+    labels_arr = measure.label(img_arr, connectivity=connectivity)
```
[the remark this answers is not in the document]
Comment (YuanTingHsieh) [reply monai/transforms/utils.py:410]: how to do this perf test, do we have an example of this somewhere?
And what is the timing we are expecting for it to be too slow?

[the remark this answers is not in the document]
Comment (Nic-Ma) [reply monai/transforms/utils.py:410]: for the first version, you can just try to run with a data with a similar size to the output of the spleen dataset. Just make sure it's not blocked(_even slower than the pre-processing + model forward()_).
Thanks.

[the remark this answers is not in the document]
Comment (YuanTingHsieh) [reply monai/transforms/utils.py:410]: With unet_evaluation_array.py
On GPU, the infer time is 0.06 seconds ( not adding pre-processing)
On CPU the infer time is 3.8 seconds
This transform is taking around 0.08 seconds for each image.
And 90% of the time is spent in "measure.label" function
I think we can't control that part for now

[the remark this answers is not in the document]
Comment (Nic-Ma) [reply monai/transforms/utils.py:410]: Great test results!
Looks good to go ahead now.

**Comment by YuanTingHsieh** [comment]
Ah I have to fix the unit test that use cuda tensor...

**Comment by Nic-Ma** [comment]
> Ah I have to fix the unit test that use cuda tensor...

You can run the unit tests locally before submitting to save time:
```
cd MONAI
bash runtests.sh --coverage
```

Thanks.

File: monai/transforms/post/dictionary.py (hunk 10 above)
Comment (YuanTingHsieh) [reply monai/transforms/post/dictionary.py]: done

**Comment by YuanTingHsieh** [comment]
I fixed the tests

File: tests/test_keep_largest_connected_component.py [hunk 12]
```diff
@@ -0,0 +1,145 @@
...
+
+TEST_CASE_11 = [
+    "value_0_background_3",
+    {"independent": False, "applied_values": [0], "background": 3},
+    grid_1,
+    torch.tensor([[[[3, 3, 1, 3, 3], [3, 2, 1, 1, 1], [1, 2, 1, 0, 0], [1, 2, 0, 1, 0], [2, 2, 0, 0, 2]]]]),
+]
+
+TEST_CASE_12 = [
+    "all_0_batch_2_cuda",
```
Comment (Nic-Ma) [inline tests/test_keep_largest_connected_component.py:100]: I think the "cuda" here is some typo, let's fix it in next PR.

File: monai/transforms/post/array.py [hunk 13]
```diff
@@ -46,3 +49,94 @@ def __call__(self, img, to_onehot=None, num_classes=None):
...
+    def __call__(self, img):
+        """
+        Args:
+            img: shape must be (batch_size, 1, spatial_dim1[, spatial_dim2, ...]).
+
+        Returns:
+            A PyTorch Tensor with shape (batch_size, 1, spatial_dim1[, spatial_dim2, ...]).
+        """
+        channel_dim = 1
+        if img.size()[channel_dim] == 1:
```
Comment (Nic-Ma) [inline monai/transforms/post/array.py:125]: `img.shape[channel_dim]` is more common.

File: monai/transforms/post/dictionary.py [hunk 14]
```diff
@@ -56,4 +56,43 @@ def __call__(self, data):
...
+                The pixel whose value is not in this list will remain unchanged.
+            independent (bool): consider several labels as a whole or independent, default is `True`.
+                Example use case would be segment label 1 is liver and label 2 is liver tumor, in that case
+                you want this "independent" to be specified as False.
+            background (int): Background pixel value. The over-segmented pixels will be set as this value.
+            connectivity (int): Maximum number of orthogonal hops to consider a pixel/voxel as a neighbor.
+                Accepted values are ranging from  1 to input.ndim. If ``None``, a full
+                connectivity of ``input.ndim`` is used.
+            output_postfix (str): the postfix string to construct keys to store converted data.
+                for example: if the keys of input data is `pred` and `label`, output_postfix is `metrics`,
```
Comment (Nic-Ma) [inline monai/transforms/post/dictionary.py:81]: maybe we can change to "output_postfix is `largest`"

File: monai/transforms/post/array.py [hunk 15]
```diff
@@ -46,3 +49,94 @@ def __call__(self, img, to_onehot=None, num_classes=None):
...
+                Accepted values are ranging from  1 to input.ndim. If ``None``, a full
+                connectivity of ``input.ndim`` is used.
+        """
+        super().__init__()
+        self.applied_values = applied_values
+        self.independent = independent
+        self.background = background
+        self.connectivity = connectivity
+        if background in applied_values:
+            raise ValueError("Background pixel can't be in applied_values")
```
Comment (Nic-Ma) [inline monai/transforms/post/array.py:114]: missed a "." at the end of this warning string.
